### PR TITLE
feature(smart-contracts): Gitcoin Passport hook

### LINKED ISSUE #13208: Gitcoin Passport hook
Can we create a Hook for `onKeyPurchase` that would require the recipient of the NFT to have a Gitcoin passport?

### PR DESCRIPTION
**Pull request by 0xTxbi**
# Description
This PR introduces a Gitcoin Passport verification hook to the onKeyPurchase function. The addition serves to restrict NFT purchases to users holding a valid Gitcoin Passport.

# Issues

Fixes #13208
Refs #13208

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [x] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by julien51** [comment]
Thanks a lot @0xTxbi ! Please add yourself to the CLA bot config file!

File: .clabot [hunk 1]
```diff
@@ -58,7 +58,8 @@
     "sudheerDev",
     "SVell",
     "JoaoCampos89",
-    "jchanolm"
+    "jchanolm",
+    "0xTxbi"
```
Comment (julien51) [inline .clabot:62]: @0xTxbi Please move this change to another PR!

Comment (julien51) [reply .clabot:62]: We can't do it for you because of course that would not mean you agree to it :)

Comment (0xTxbi) [reply .clabot:62]: very true. done. moved it to another PR here: #13432

File: smart-contracts/contracts/hooks/GitcoinHook.sol [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    address recipient,
+    address /* referrer */,
+    bytes calldata signature /* data */
+  ) external view returns (uint256 minKeyPrice) {
+    string memory message = toString(recipient);
+    require(checkIsSigner(message, signature), "WRONG_SIGNATURE");
+    if (address(msg.sender).code.length > 0) {
+      return IPublicLockV13(msg.sender).keyPrice();
+    }
+    return 0;
```
Comment (julien51) [inline smart-contracts/contracts/hooks/GitcoinHook.sol:35]: We should probably never return `0` here because that means the user could pay `0` instead.
I think a safer approach is just to revert here?

**Review by julien51 — CHANGES_REQUESTED** [review]
Let's make sure we do NOT return `0` in any case.
Also I would love to have @clemsos check this!

File: smart-contracts/contracts/hooks/GitcoinHook.sol (hunk 2 above)
Comment (0xTxbi) [reply smart-contracts/contracts/hooks/GitcoinHook.sol]: you're right, that was an oversight. great observation! now fixed!

**Review by julien51 — APPROVED** [review]
